File: mobile_client/app/src/main/java/wpam/mobile_client/client/ResponseHandlerVisitor.java
package wpam.mobile_client.client;


import android.content.Context;
import android.content.Intent;
import android.support.v4.content.LocalBroadcastManager;

import wpam.mobile_client.MainActivityInterface;
import wpam.mobile_client.protocol.*;
import wpam.mobile_client.sensor_tag.PlotInterface;
import wpam.mobile_client.sensor_tag.SensorsInterface;

import java.io.*;
import java.text.SimpleDateFormat;
import java.util.Date;


public final class ResponseHandlerVisitor extends ResponseVisitor implements Closeable
	{
		private MainActivityInterface mainActivityInterface;
		private Context context;
		private PlotInterface plotInterface;

		SimpleDateFormat sdf = new SimpleDateFormat("yyyy.MM.dd HH:mm:ss");

		String blindsStatus;

		ClientThread client;

		public ResponseHandlerVisitor(ClientThread client)
		{
			this.client = client;
		}
		public final void close()
		{
		}

		public void setMainActivityInterface(MainActivityInterface mainActivityInterface)
		{
			this.mainActivityInterface = mainActivityInterface;
		}

		public void setSensorsTagInterface(Context context)
		{
			this.context = context;
		}

		public void setPlotInterface(PlotInterface plotInterface)
		{
			this.plotInterface = plotInterface;
		}

		@Override
		public  void visit(AuthenticateResponse response)
		{
			System.out.println(response.getName());
			String currentDateandTime = sdf.format(new Date());
			client.addToQueue("--" + currentDateandTime +"-- Received:: " + response.getName() + ".");
		}
		
		@Override
		public  void visit(AckResponse response)
		{
			String currentDateandTime = sdf.format(new Date());
			client.addToQueue("--" + currentDateandTime +"-- Received:: "+
					response.getName() + ", type of ack: " + response.getAckType().toString() + ".");
			this.mainActivityInterface.thiefOccurred();
		}
		
		@Override
		public  void visit(ErrorResponse response)
		{
			String currentDateandTime = sdf.format(new Date());
			client.addToQueue("--" + currentDateandTime +"-- Received:: " +
					response.getName() + ", type of error: " + response.getErrorType().toString() + ".");
			System.out.println(response.getName());
		}

		@Override
		public  void visit(MotorStatusResponse response)
		{
			String currentDateandTime = sdf.format(new Date());
			client.addToQueue("--" + currentDateandTime +"-- Received:: " +
					response.getName() + ", motor status: " + response.getMotorStatus().toString() + ".");
			System.out.println(response.getName());
		}

		@Override
		public  void visit(BlindsStatusResponse response)
		{
			String currentDateandTime = sdf.format(new Date());
			client.addToQueue("--" + currentDateandTime +"-- Received:: " +
					response.getName() + ", blinds status: " + response.getBlindsStatus().toString() + ".");


			BlindsStatus status = response.getBlindsStatus();
			blindsStatus = "Blinds status: " + status.name();

			switch (status)
			{
				case UP:
				{
					mainActivityInterface.blindsUP();
					break;
				}
				case DOWN:
				{
					mainActivityInterface.blindsDOWN();
					break;
				}
			}

		}

		@Override
		public  void visit(GuardStatusResponse response)
		{
			String currentDateandTime = sdf.format(new Date());
			client.addToQueue("--" + currentDateandTime +"-- Received:: " +
					response.getName() + ", guard status: " + response.getGuardStatus().toString() + ".");

			if(response.getGuardStatus() == GuardStatus.OFF)
			{
				client.setUserInHome(true);
			}
			else
			{
				client.setUserInHome(false);
			}
			System.out.println(response.getName() + " " + response.getGuardStatus().toString());
		}

		@Override
		public void visit(PlotResponse data) {
			String currentDateandTime = sdf.format(new Date());
			client.addToQueue("--" + currentDateandTime +"-- Received:: " +
					data.getName() + ".");

			plotInterface.plotData(data.getValues(), data.getTimestamps());
		}

		@Override
		public void visit(SensorTagSamplesResponse response)
		{
			String currentDateandTime = sdf.format(new Date());
			client.addToQueue("--" + currentDateandTime +"-- Received:: " +
					response.getName() + ".");

			Intent incomingMsg = new Intent("SensorTagSamplesResponse");
			incomingMsg.putIntegerArrayListExtra("data", response.getValues());

			LocalBroadcastManager.getInstance(this.context).sendBroadcast(incomingMsg);


			//sensorsInterface.singleSamples(response.getValues());
		}

		public String getBlindsStatus()
		{
			return blindsStatus;
		}
	}

File: server/src/main_client/client_tcp/ResponseHandlerVisitor.cpp
#include "ResponseHandlerVisitor.h"

using namespace std;
using namespace communication;
using namespace utility;

ResponseHandlerVisitor::ResponseHandlerVisitor()
    : logger_(Logger::getInstance())
{}

ResponseHandlerVisitor::~ResponseHandlerVisitor()
{}

void ResponseHandlerVisitor::visit(AckResponse &data)
{
    if(logger_.isInformationEnable())
    {
        string message;
        if (data.getAckType() == AckType::OK) {
            message = string("ResponseHandlerVisitor :: Received AckResponse, ackTYPE: OK");
        } else {
            message = string("ResponseHandlerVisitor :: Received AckResponse, ackTYPE: FAIL");
        }

        logger_.writeLog(LogType::INFORMATION_LOG, message);
    }
}

void ResponseHandlerVisitor::visit(DataResponse &data)
{
    if(logger_.isInformationEnable())
    {
        const string message = string("ResponseHandlerVisitor :: Received DataResponse, data: ") + data.getData();
        logger_.writeLog(LogType::INFORMATION_LOG, message);
    }
}

void ResponseHandlerVisitor::visit(ErrorResponse &data)
{
    if(logger_.isInformationEnable())
    {
        const string message = string("ResponseHandlerVisitor :: Received DataResponse, data: ") + to_string(data.getErrorType());
        logger_.writeLog(LogType::INFORMATION_LOG, message);
    }
}

void ResponseHandlerVisitor::visit(MotorStatusResponse &data)
{
    if(logger_.isInformationEnable())
    {
        const string message = string("ResponseHandlerVisitor :: Received DataResponse, data: ") + to_string(data.getMotorStatus());
        logger_.writeLog(LogType::INFORMATION_LOG, message);
    }
}

void ResponseHandlerVisitor::visit(GuardStatusResponse &data)
{
    if(logger_.isInformationEnable())
    {
        const string message = string("ResponseHandlerVisitor :: Received DataResponse, data: ") + to_string(data.getGuardStatus());
        logger_.writeLog(LogType::INFORMATION_LOG, message);
    }
}

void ResponseHandlerVisitor::visit(BlindsStatusResponse &data)
{
    if(logger_.isInformationEnable())
    {
        const string message = string("ResponseHandlerVisitor :: Received DataResponse, data: ") + to_string(data.getBlindsStatus());
        logger_.writeLog(LogType::INFORMATION_LOG, message);
    }
}

void ResponseHandlerVisitor::visit(AuthenticateResponse& data)
{
    if(logger_.isInformationEnable())
    {
        const string message = string("ResponseHandlerVisitor :: Received Authentication message.");
        logger_.writeLog(LogType::INFORMATION_LOG, message);
    }
}

void ResponseHandlerVisitor::visit(SensorTagSamplesResponse& data)
{
    if(logger_.isInformationEnable())
    {
        const string message = string("ResponseHandlerVisitor :: Received SensorTagSamplesResponse message.");
        logger_.writeLog(LogType::INFORMATION_LOG, message);
    }
}

void ResponseHandlerVisitor::visit(PlotResponse& data)
{
    if(logger_.isInformationEnable())
    {
        const string message = string("ResponseHandlerVisitor :: Received SensorTagSamplesResponse message.");
        logger_.writeLog(LogType::INFORMATION_LOG, message);
    }
}